README.md
IceBlock
========

Simplified example shooting beam of particles into giant piece of ice.

In order to run, you will need to update setupfiles/setup such that
the environment variables are set specifically for your machine. This
means that you need to source the geant4make.sh script in your 
geant4-build directory.  You also need to make sure that the data
files containing cross-sectional info are pointing in the right 
area.  My $G4INSTALL points to geant4, not geant4-build, so I 
additionally had to link the data/ directory in geant4-build to 
my geant4 directory like so:

cd /path/to/untarred-geant4/
ln -s ../geant4-build/data data

To execute: 
iceblock -ne <number_of_events> -e <beam_energy> -np <number_of_primaries>

<number_of_events> = number of events specified as an int
<beam_energy> = energy of the particle beam specified as an int
<enum_for_material_type> = an int for material type:
			 - 0 = ice
			 - 1 = lead
			 - 2 = iron

You can get more info from running:
iceblock -h

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=                      UPDATES AS OF AUGUST 2015                        =
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Since this README was written we have made the following additions:

1.) Beam Shape -- we now allow one to set the beam shape to either:
    	 * Single injection point (default)
	 * 2-D Gaussian with some width by using: --gauss 45 (for 45mm)
	 * Flat uniform distribution by using: --flat 45 (squre with dimension 45mm)

2.) Bunch Number -- This will allow one to specify the number of bunches. Please
    note that when this option is used, it creates bunches with the same number
    of primaries specified at np.  So for example -np 1000 --nbunch 5 will create
    5 bunches with 1000 primaries

3.) Timing Offset -- This allows one to control the timing difference between 
    bunches in the bunch train.  The default is 350ps, as the beam in Utah

4.) Beam Profile -- We allow one to specify a beam profile.  What that means is we
    allow one to imput a more complicated bunch structure instead of equal bunches.
    This allows control of the relative particle number in each bunch.  One thing
    to note: The total number of particles set using -np is distributed over the
    input bunch structure.  A simple case of 3 bunches, where 1st has 10% of 
    the charge, and 2nd has 50% and 3rd has 40%.  So for -np 100 you would 
    get: 10, 50, and 40 particles in the 1st, 2nd, and 3rd bunches respectively.

5.) Step Limit -- This will force Geant4 to take a step even time, regardless 
    if any physics interaction is taking place.  This is useful for the Endpoint
    method, I believe, and that is why it was added.  ZHS doesn't care about it.

6.) Antenna file -- The user can specify a text file for the Antenna positions.  
    Unfortunately I haven't had time to clean this up, but in reality the only 
    important numbers are x, y, z.  This should be cleaned up in the future
    and potentially merged with the lookup table inputs.


#-------------------------------------------------------------#
# Current Issues that could be cleaned up or dealt with
#-------------------------------------------------------------#

1.) Right now there is no way to easily run without refraction and with 
    refraction turned on.

2.) We may need a better way to handle lookup tables, or at least
    generate lookup tables for finer angular spacings. This will
    be important for characterizing in simulation the angular
    dependence of TR and Askaryan in our Delta Ice

3.) Maybe think of adding a command line option for changing index of 
    refraction.  This might be a source of systematic error in the end.

4.) We need a robust way to test the lookup tables.  Some parameters
    or code to verify that they are ok to be used for both the outside
    points and the inside points.


Changelog
2015-08-11 Matt:
	* Moved geometry from air --> ice
	* Turned off some of the printing

2015-08-05 Matt:
	* Creating branch to study transition radiation
	* Allowing energy to take in float from command line to study 3.6 MeV as done in Paper
	* Moving ice to be above 0 and air below 0 in z direction
	* Adding TRFromZHS method to Stepping Action
	* Committing Constants.hh and the TRTool with working copy
	* Geometry is ice --> air!!!

2015-02-24 Matt:
	* Trying to clean up a little.  First round done.

2015-02-10 Matt:
	* Adding new class BeamProfile.hh to read in text file from profile.
	* Charge profile structure included for the beam.
	* Data directory added for necessary data file storage
	* Beam charge distribution added to data

2015-02-09 Matt:
	* Checked update of increasing beam width
	* Need to incorporate beam structure in next update

2014-11-18 Matt:
	* Bunch structure added with delta function spikes.
		- Will add more later to test this crazy bunch structure we will have

2014-11-06 Matt:
	* There was an error in how I was handling the bins. I was putting things into a Profile,
		which will average the result if multiple entries end up in one bin. However this
		is not the way to do it. We are evaluating a function a specific steps in time,
		therefore the binning must remain fixed.  It is too hard to do with scattered
		root scripts, therefore I have written new analysis code.
	* Updates to all scripts that were used in testing are now committed
	* Tagging and then will introduce bunch structure to the beam.

2014-10-17 Matt:
	* Fixing time check in SteppingAction. Including everything in window.
	* Moving beam slightly off-center for all particles in one position case
	* Minor updates to python scripts.
	* Adding miscellanous plotting scripts
	* Tagging before testing some updates

2014-10-02 Matt:
	* Updating EField and APotPlots. Latest investigations have been added.
	* Tagging Version 00.04
	* Added options --flat <float> and --gauss <float> such that beam profile can be set
	* Adding short script to plot beam profile from output: PlotBeamProfile
		This script needs output from PrimaryGeneratorAction. Search f_test.
	* Adding some methods to myHist.C
	* Fixing PrimaryGeneratorAction. Wasn't looping over n_particles.
	* Adding plotting script to compare the point source, flat, and gaussian particle guns.

2014-10-01 Matt:
	* Including corrected makeRefractedScan.py (forgot previously). Finished now
	* Updating Antenna.{cc,hh} to store angle information
	* Updating SetupAntenna to handle angle information
	* Updating EventAction to write the angle information to txt file
	* Updating refractive scane and everything mentioned above to handle calculation of the
		refracted electric field. Previous method was incorrect approach.
	* Updating EField calculation for the comparison refracted with unrefracted
	* Adding refractedAngles script to plot refracted angles vs. height

2014-09-30 Matt:
	* Changed antenna position for default to 1km for comparison with ZHS.
	* Checked agreement between ZHS and G4 for 40 MeV after bug fixes.
	* Added all plots to output root file from APotPlots.py
	* updated EField script to check scaling results. Everything seems consistent
	* Updated VPCompareZHS to include latest ZHS comparison files.
	* Updated iceblock from R10m to R1000m following update in antenna position
	* TODO: Add antennta default distance as a parameter
	* Adding makeRefractionScan.py -- This is still work in progress.
		- Right now for angles near critical angle, z increases when it should continue
		decreasing. I think it has to do with the sign ambiguity. Try to impose some
		requirements.

2014-09-29 Matt:
	* Updating some minor changes after figuring out difference with ZHS.
	* Committing verified version with velocity correction in place.
	* Committing E-field drawing script.  Need to add more modularity for future

2014-09-09 Matt:
	* Committing debugged vector potential calculation. Right now two ways to calculate the velocity.
	* Adding python script to make Vector potential plots
	* Adding python script to make Cherenkov Ring plots
	* Adding python script to make filelists for Zscan and Cherenkov scan
	* Adding scripts to compare velocity calculation procedure
	* Adding scripts to format cherenkov scan results
	* Adding script to format the vector potential comparison for ZHS vs. Geant4
	* Adding class for Antenna and one for setting up the antenna


2014-08-5 Matt:
	* Adding vector potential calculation. First pass looks reasonable.
	* Update SetupAntenna to loop over angles for fixed position.
	* Fixed timing bin factor in calculation of vector potential

2014-08-04 Matt:
	* Minor updates to settings.
	* Going to add calculation of electric field for step information. Tag first

2014-05-29 Matt:
	* Updating some minor changes during running.
	* Going to build realistic setup for ELS site.

2014-03-27 Matt:
	* Add plotting macro for shower comparisons
	* Update ntuple maker to account for nParticles

2014-03-25 Matt:
	* Add option to specify the number of initial particles.

2014-03-19 Matt:
	* Added option to set tracking threshold
	* Added script to make simple tree.  We will see if it is useful
	* Committing fixed version of code to make simple tree

2014-03-13 Matt:
	* Adding python scripts used to make all the plots
	* Adding C scripts for formatting plot output
	* Adding submission scripts to give idea of how to run

2014-03-12 Matt:
	* TTree takes too long! Going back to text files.
	* Added modularity to save both track and step output in separate files.

2014-03-10 Matt:
	* After many changes, finally figured out my issue with particle counting. Commit changes before next update.
	* Next Update will be to create a TTree to be output of Geant information.

2014-02-14 Matt:
	* Committing revisions and current stable working version.
	* After trying many things to debug, I am pausing and going to tag this version

iceblock.cc


//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-//
// Full Simulation using Geant4 to calculate the electric field for an //
// EM shower in Ice.  It is to be used for the ARA TA-ELS experiment   //
// conducted in Utah.                                                  // 
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-//


// Geant4 Classes
#include "G4RunManager.hh"
#include "G4UImanager.hh"
#include "DetectorConstruction.hh"
#include "PhysicsList.hh"
#include "PrimaryGeneratorAction.hh"
#include "RunAction.hh"
#include "EventAction.hh"
#include "SteppingAction.hh"
#include "SteppingVerbose.hh"
#include "TrackingAction.hh"

// Some additional classes for convenience
#include "BeamProfile.hh"
#include "SetupAntenna.hh"
#include "RefractionTool.hh"

// Standard
#include <sstream>
#include <fstream>
#include "time.h"

//---------------------------------------------------//
// Help menu for all the options
//---------------------------------------------------//
void help()
{

  cout << endl;
  cout << endl;
  cout << "------------------------------------------------------------" << endl;
  cout << "-ne <int> " << endl;
  cout << "\t Specify the number of events " << endl;
  cout << "\t default is 1" << endl;
  cout << "-np <int> " << endl;
  cout << "\t Specify the number of particles " << endl;
  cout << "\t default is 1 " << endl;
  cout << "-e <float> " << endl;
  cout << "\t Specify the beam energy in MeV" << endl;
  cout << "\t Default is 1000 MeV" << endl;
  cout << "-t <int> " << endl; 
  cout << "\t Specify the interaction material " << endl;
  cout << "\t Default is 0 (ice) " << endl;
  cout << "\t\t 0 -- ice " << endl;
  cout << "\t\t 1 -- lead " << endl;
  cout << "\t\t 2 -- iron " << endl;
  cout << "-p <char> " << endl;
  cout << "\t Specify the particle type" << endl;
  cout << "\t Default is e-" << endl;
  cout << "\t\t e- -- Electron" << endl;
  cout << "\t\t e+ -- Positron" << endl;
  cout << "\t\t mu- -- Muon" << endl;
  cout << "\t\t gamma -- Photon" << endl;
  cout << "-c <float>" << endl;
  cout << "\t Specify the energy threshold to use in MeV" << endl;
  cout << "\t Default is 0" << endl;
  cout << "-i <file.txt> " << endl;
  cout << "\t Defaule is NULL" << endl;
  cout << "\t\tSpecify input file to read antenna config from"<<endl;
  cout << "--flat <float>" << endl;
  cout << "\t Inject random flat distribution with width <float> [mm]" << endl;
  cout << "--gauss <float>" << endl;
  cout << "\t Inject Gaussian distribution with <float> sigma [mm]" << endl;
  cout << "--nbunch <int>" << endl;
  cout << "\t Create N bunches" << endl;
  cout << "--offset <float>" << endl;
  cout << "\t Timing offset for bunches (default 0.35 ns)"<<endl;
  cout << "--beam <file.txt> " << endl;
  cout << "\t Input beam profile file" << endl;
  cout << "--stepLimit <int> " << endl;
  cout << "\t Input step limit in mm (default is -1, off)" << endl;
  cout << "--refCoeff" << endl;
  cout << "\t Include Fresnel coefficients" << endl;
  cout << "--rot <int>" << endl;
  cout << "\t Specify rotation angle. Default is 30." << endl;
  cout << "------------------------------------------------------------" << endl;
  cout << endl;
  cout << endl;

}

//---------------------------------------------------//
// Main
//---------------------------------------------------//
int main(int argc, char** argv)
{

  // Store time
  clock_t tStart = clock();

  //============================================//
  // List all the options here with some small
  // description of what they are used for
  //============================================//
  G4int nEvents          = 1;     // number of events
  G4int nParticles       = 1;     // number of particles
  G4float beamEnergy     = 1000;  // beam energy
  G4int detMaterial      = 0;     // detector material
  std::string partType   = "e-";  // primary particle
  G4double threshold     = 0;     // energy threshold for testing -- leave 0
  bool useThreshold      = false; // Whether or not to use this threshold -- don't
  std::string antFile    = "";    // File path for antenna positions
  G4bool b_flat          = false; // Use uniform beam
  G4bool b_gauss         = false; // Use Gaussian distributed beam
  G4double sigma         = 0;     // width of the beam
  G4int nbunches         = 1;     // Number of bunches in beam
  G4double tOffset       = 0.350; // Timing between bunches [ns]
  std::string beamFile   = "";    // Beam profile from txt file
  G4double stepLimit     = -1;    // Beam step limit
  G4bool useRefTool      = false; // Turn on refraction
  G4int theSeed          = 564738291; // Generic seed
  G4int rotation         = 30;        // Degree

  //============================================//
  // Load the options
  //============================================//
  for(int i=1; i<argc; ++i){
    if( strcmp(argv[i], "-ne") == 0 )
      nEvents = atoi( argv[++i] );
    else if( strcmp(argv[i], "-np") == 0 )
      nParticles = atoi( argv[++i] );
    else if( strcmp(argv[i], "-e") == 0 )
      beamEnergy = atof( argv[++i] );
    else if( strcmp(argv[i], "-t") == 0 )
      detMaterial = atoi( argv[++i] );
    else if( strcmp(argv[i], "-p") == 0 )
      partType = string(argv[++i]);
    else if( strcmp(argv[i], "-c") == 0 ){
      threshold = atof( argv[++i] );
      useThreshold = true;
    }
    else if( strcmp(argv[i], "-i") == 0 )
      antFile = argv[++i];
    else if( strcmp(argv[i], "--flat") == 0 ){
      b_flat = true;
      sigma = atof( argv[++i] );
    }
    else if( strcmp(argv[i], "--gauss") == 0 ){
      b_gauss = true;
      sigma = atof( argv[++i] );
    }
    else if( strcmp(argv[i], "--nbunch") == 0 )
      nbunches = atoi( argv[++i] );
    else if( strcmp(argv[i], "--offset") == 0 )
      tOffset = atof( argv[++i] );
    else if( strcmp(argv[i], "--beam") == 0 )
      beamFile = argv[++i];
    else if( strcmp(argv[i], "--stepLimit") == 0)
      stepLimit = atof( argv[++i] );
    else if( strcmp(argv[i], "--refCoeff") == 0)
      useRefTool = true;
    else if( strcmp(argv[i], "--seed" ) == 0 )
      theSeed = atoi( argv[++i] );
    else if( strcmp(argv[i], "--rot" ) == 0 )
      rotation = atoi( argv[++i] );
    else{
      help();
      return 0;
    }
  }//end loop over arguments

  // Make sure det material makes sense
  if( detMaterial > 2 ){
    cout<<"Error: Det material not recognized: "<<detMaterial<<endl;
    help();
    return 0;
  }

  //============================================//
  // Make an output name including the run 
  // information. Perhaps could come up with 
  // something better.
  // THIS IS GETTING TOO LONG
  //============================================//
  stringstream ss;
  ss << "output_" << nEvents << "_" << beamEnergy << "_";

  // Check det material
  if(detMaterial == 1)      ss << "lead";
  else if(detMaterial == 2) ss << "iron";
  else                      ss << "ice";

  // Check particle type
  if(partType == "e-")         ss << "_eBeam";    
  else if(partType == "mu-")   ss << "_muBeam";    
  else if(partType == "e+")    ss << "_pBeam";    
  else if(partType == "gamma") ss << "_gBeam";
  else                         ss << "_unkBeam";
  
  // Add the number of particles
  ss << "_np" << nParticles;

  // Decide if antenna file specified
  if( !antFile.empty() ){
    string s_ant = antFile.substr(0,antFile.find(".txt"));
    s_ant        = s_ant.substr(antFile.find("/")+1,-1);
    cout<<s_ant<<endl;
    ss << "_" << s_ant;
  }
  else{ // uses a single antenna a theta_c 100m away
    ss << "_HardCodedAntenna_R100m";
  }

  // Add some more info about input distribution
  BeamProfile* bp = new BeamProfile();
  if(b_flat)  ss << "_RandFlat" << sigma;
  else if(b_gauss) ss << "_RandGauss" << sigma;
  else             ss << "_singlePos";

  // If threshold is set, append to file
  if(useThreshold) ss << "_thresh" << threshold << "MeV";

  // Add number of bunches information
  if(beamFile.empty()) ss << "_bunches" << nbunches;
  else{
    bp->init(beamFile);
    if( !bp->isInit() ){
      cout<<"************************************"<<endl;
      cout<<"Error: Beam Profile not initialized"<<endl;
      cout<<"************************************"<<endl;
      return 0.;
    }
    string s_beam = beamFile.substr(0,beamFile.find(".txt"));
    s_beam        = s_beam.substr(beamFile.find("/")+1,-1);
    cout<<"Loading beam: "<<s_beam<<endl;
    ss << "_" << s_beam;
    //nbunches = bp->getN(); 
  }

  // Add beam information
  cout<<"Step limit: "<<stepLimit<<endl;
  if( stepLimit > 0 ){
    ss << "_steplimit" << stepLimit; 
  }

  // Create the refraction tool
  RefractionTool* refTool = new RefractionTool();
  refTool->setUse(useRefTool);
  if(useRefTool) ss << "_refractionOn";

  // Add icetilt to fname
  ss << "_tilt" << rotation << "deg";

  // New feature: choose seed
  ss << "_seed" << theSeed; 

  //ss << "_testingLookUp";
  //ss << "_testingshiz_updatedLookup";
  ss << "_TRTEST";

  //============================================//
  // Configure Geant Below
  //============================================//

  // Setup the Antennas
  SetupAntenna* m_AntSetup = new SetupAntenna(antFile);
  std::vector<Antenna*> m_Ants = m_AntSetup->getAnts();  

  // Default run manager
  G4RunManager* runManager = new G4RunManager;

  // Construct detector
  DetectorConstruction* detector = new DetectorConstruction(detMaterial,
							    threshold,
							    useThreshold,
							    stepLimit,
							    refTool,
							    &m_Ants,
							    rotation);
  
  runManager->SetUserInitialization(detector);

  // Set Physics list
  G4VUserPhysicsList* physics = new PhysicsList(useThreshold);
  runManager->SetUserInitialization(physics);
  
  // Open up a file for output
  // This is not currently used anymore since we have the
  // e-field directly calculated
  //std::ofstream trackOutput(("tracks/"+ss.str()+".dat").c_str(), std::ofstream::out);
  //std::ofstream stepOutput(("steps/"+ss.str()+".dat").c_str(), std::ofstream::out);
  std::ofstream APotOut(("efield/A_"+ss.str()+".dat").c_str(), std::ofstream::out);
  std::ofstream EFOut(("efield/E_"+ss.str()+".dat").c_str(), std::ofstream::out);

  // Output some meta data info for the vector potential plots
  APotOut << "# " << nEvents << " " 
	  << nParticles << " "
	  << beamEnergy << " "
	  << m_Ants.size() << " "
	  << m_Ants.at(0)->getNPoints() << " "
	  << m_Ants.at(0)->getTStep()   << " "
	  << nbunches                   << " "
	  << tOffset
	  << G4endl;

  EFOut << "# " << nEvents << " " 
	<< nParticles << " "
	<< beamEnergy << " "
	<< m_Ants.size() << " "
	<< m_Ants.at(0)->getNPoints() << " "
	<< m_Ants.at(0)->getTStep()   << " "
	<< nbunches                   << " "
	<< tOffset
	<< G4endl;

  // Set Primary action generator which will inject particles
  PrimaryGeneratorAction* genAction = new PrimaryGeneratorAction(detector,
								 beamEnergy,
								 partType,
								 nParticles,
								 b_flat, b_gauss, sigma,
								 nbunches, tOffset,
								 bp, theSeed);  
  
  runManager->SetUserAction(genAction);

  // Setup the actions to carry out  
  runManager->SetUserAction(new RunAction(detector,genAction));
  runManager->SetUserAction(new EventAction(NULL, NULL, //&trackOutput, &stepOutput,
					    &APotOut,
					    &EFOut,
					    &m_Ants));
  runManager->SetUserAction(new SteppingAction(NULL /*&stepOutput*/, &m_Ants, refTool));
  //runManager->SetUserAction(new TrackingAction(&trackOutput));

  // Initialize G4 Kernel
  runManager->Initialize();

  // Start the run
  runManager->BeamOn(nEvents);

  // Close up the files
  //trackOutput.close();
  //stepOutput.close();
  APotOut.close();
  EFOut.close();
    
  // Dump some info
  cout<<"Number of particles "<<nParticles<<endl;
  cout<<"Runtime is: "<<(double)(clock() - tStart)/CLOCKS_PER_SEC<<" seconds "<<endl;

  // Clean up and finish
  delete m_AntSetup;
  delete runManager;
  delete refTool;
  return 0;

}
